docs/IntegrationTest.md

```diff
@@ -0,0 +1,191 @@
+# Integration Tests
+
+## Jenkins on GKE with helm
+
+[Start Cloud Shell](https://cloud.google.com/shell/docs/quickstart#start_cloud_shell) from the
+GCP console. All commands listed below are to be run from the Cloud Shell CLI
+
+### Google Cloud Platform Prerequisites
+1. If you are performing these steps in a different CLI session than the one you used for the
+[usage](Home.md#usage) tutorial then make sure the `$PROJECT` environment variable is exported:
+    ```bash
+    export PROJECT=$(gcloud info --format='value(config.project)')
+    ```
+
+1. Make sure that `kubectl` is installed:
+    ```bash
+    kubectl version
+    ```
+
+    You can install through the [Google Cloud SDK](
+    https://kubernetes.io/docs/tasks/tools/install-kubectl/#download-as-part-of-the-google-cloud-sdk)
+    (which should also be installed on Cloud Shell):    
+    ```bash
+    gcloud components install kubectl
+    ```
+
+    If not, choose one of the other installation methods
+    [here](https://kubernetes.io/docs/tasks/tools/install-kubectl/).
+
+1. Create a GKE cluster similar to step 1 of [Configure GKE Cluster](Home.md#configure-gke-cluster):
+    ```bash
+    export CLUSTER=jenkins-master
+    export ZONE=us-central1-c
+    gcloud container clusters create --zone $ZONE $CLUSTER
+    ```
+
+1. Create a [domain](https://domains.google/) for hosting your jenkins instance. Export that now:
```

Comment:
The customer might have an existing domain, just for clarity might want to add some prose about being able to use an existing domain.

Reply:
Good catch, I'll change this

Reply:
If the goal is just to run the integration tests, you could eschew externalizing the Jenkins instance and instead connect to it via a port-forward. For example step 3 here:
https://cloud.google.com/solutions/jenkins-on-kubernetes-engine-tutorial#installing_jenkins

Relatedly, why not point to that doc then start this doc at the next section?

Reply:
I agree that it would be better not to externalize the Jenkins instance, but I'd want to be able to use a secure connection still for uploading the credentials, is there an easy way to modify the existing `jenkins/values.yaml` file or the port forward to enable that in the short term?

There are a few other things missing from that tutorial's `jenkins\values.yaml` such as the custom agent image and some outdated/missing plugins like the 0.6 google-oauth-plugin vs the required 0.7, but it would be simple enough to say "use this values.yaml instead of the one that the tutorial provides".

Most of this extra set-up revolves around being able to create an ingress and connect to it externally with https, so I'd be able to replace a lot of this with a link to that documentation if that requirement was removed. Another thing that would enable cleaning this up a great deal would be the refactor to the google-oauth-plugin so that we can use the CasC values to include the credentials in the helm chart and that way the key never leaves and doesn't have to be uploaded over the web to GCP, so in addition to avoiding the need to link it externally, we could just use the default port forwarding setup already listed in that doc.

Reply:
A port-forwarded connection is actually encrypted across to the Kubernetes Master.

I'd be happy to make the changes to the upstream solution doc and repo as needed so that we dont duplicate the content here.

Reply:
First part here:
https://github.com/GoogleCloudPlatform/continuous-deployment-on-kubernetes/pull/144

For the custom image, have we decided on a place to host it? If so I'll add that value to the upstream solution.

Reply:
Thanks for submitting that. As for the custom image we don't know yet exactly where that would be hosted, just that it would be on us.gcr.io or gcr.io. Do you think we should use a separate project to host the image with? Maybe a project called something like gkeplugintests

Reply:
I'd just call it something like `jenkins-gke-plugin` since it's use won't be limited to tests, people may actually build off of it. The larger question would be where to host the Dockerfile. Here or in the CD on K8s repo in a folder like jenkins/agent-image:
https://github.com/GoogleCloudPlatform/continuous-deployment-on-kubernetes/tree/master/jenkins

Reply:
Now that I think about it, I don't think the custom agent image should necessarily be used in the tutorial you linked, and the docker file should be hosted here, although maybe under another directory. I think it would be okay to have this tutorial reference the tutorial that you linked, with maybe one small tweak such as:
"Use the following options in the helm install"
```
 --set Agent.Image=us.gcr.io/jenkins-gke-plugin/agent --set Agent.Priviliged=true
```

Reply:
That custom agent image and the privileged flag are the only aspects to this set-up that are specific to using this plugin for the helm chart. Everything else in our custom helm chart and set up is related to setting up an ingress which it sounds like won't be required.

Reply:
I created the project and uploaded the image which is now public: `us.gcr.io/jenkins-gke-plugin/jenkinsagent`. It has both `latest` and `lts` as tags so the helm chart default value will work.